File: Provenance/Controller/PVGBControllerViewController.m
//
//  PVGBControllerViewController.m
//  Provenance
//
//  Created by James Addyman on 22/03/2015.
//  Copyright (c) 2015 James Addyman. All rights reserved.
//

#import "PVGBControllerViewController.h"
#import <PVGB/PVGBEmulatorCore.h>

@interface PVGBControllerViewController ()

@end

@implementation PVGBControllerViewController

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    
    for (JSButton *button in [self.buttonGroup subviews])
    {
        if (![button isMemberOfClass:[JSButton class]])
        {
            continue;
        }
        
        if ([[[button titleLabel] text] isEqualToString:@"A"])
        {
            [button setTag:PVGBButtonA];
        }
        else if ([[[button titleLabel] text] isEqualToString:@"B"])
        {
            [button setTag:PVGBButtonB];
        }
    }
    
    [self.startButton setTag:PVGBButtonStart];
    [self.selectButton setTag:PVGBButtonSelect];
}

- (void)dPad:(JSDPad *)dPad didPressDirection:(JSDPadDirection)direction
{
    PVGBEmulatorCore *gbCore = (PVGBEmulatorCore *)self.emulatorCore;
    
    [gbCore releaseGBButton:PVGBButtonUp];
    [gbCore releaseGBButton:PVGBButtonDown];
    [gbCore releaseGBButton:PVGBButtonLeft];
    [gbCore releaseGBButton:PVGBButtonRight];
    
    switch (direction)
    {
        case JSDPadDirectionUpLeft:
            [gbCore pushGBButton:PVGBButtonUp];
            [gbCore pushGBButton:PVGBButtonLeft];
            break;
        case JSDPadDirectionUp:
            [gbCore pushGBButton:PVGBButtonUp];
            break;
        case JSDPadDirectionUpRight:
            [gbCore pushGBButton:PVGBButtonUp];
            [gbCore pushGBButton:PVGBButtonRight];
            break;
        case JSDPadDirectionLeft:
            [gbCore pushGBButton:PVGBButtonLeft];
            break;
        case JSDPadDirectionRight:
            [gbCore pushGBButton:PVGBButtonRight];
            break;
        case JSDPadDirectionDownLeft:
            [gbCore pushGBButton:PVGBButtonDown];
            [gbCore pushGBButton:PVGBButtonLeft];
            break;
        case JSDPadDirectionDown:
            [gbCore pushGBButton:PVGBButtonDown];
            break;
        case JSDPadDirectionDownRight:
            [gbCore pushGBButton:PVGBButtonDown];
            [gbCore pushGBButton:PVGBButtonRight];
            break;
        default:
            break;
    }
    
	[super dPad:dPad didPressDirection:direction];
}

- (void)dPadDidReleaseDirection:(JSDPad *)dPad
{
    PVGBEmulatorCore *gbCore = (PVGBEmulatorCore *)self.emulatorCore;
    
    [gbCore releaseGBButton:PVGBButtonUp];
    [gbCore releaseGBButton:PVGBButtonDown];
    [gbCore releaseGBButton:PVGBButtonLeft];
    [gbCore releaseGBButton:PVGBButtonRight];
	[super dPadDidReleaseDirection:dPad];
}

- (void)buttonPressed:(JSButton *)button
{
    PVGBEmulatorCore *gbCore = (PVGBEmulatorCore *)self.emulatorCore;
    [gbCore pushGBButton:[button tag]];
    
	[super buttonPressed:button];
}

- (void)buttonReleased:(JSButton *)button
{
    PVGBEmulatorCore *gbCore = (PVGBEmulatorCore *)self.emulatorCore;
    [gbCore releaseGBButton:[button tag]];
	[super buttonReleased:button];
}

- (void)pressStartForPlayer:(NSUInteger)player
{
    PVGBEmulatorCore *gbCore = (PVGBEmulatorCore *)self.emulatorCore;
    [gbCore pushGBButton:PVGBButtonStart];
	[super pressStartForPlayer:player];
}

- (void)releaseStartForPlayer:(NSUInteger)player
{
    PVGBEmulatorCore *gbCore = (PVGBEmulatorCore *)self.emulatorCore;
    [gbCore releaseGBButton:PVGBButtonStart];
	[super releaseStartForPlayer:player];
}

- (void)pressSelectForPlayer:(NSUInteger)player
{
    PVGBEmulatorCore *gbCore = (PVGBEmulatorCore *)self.emulatorCore;
    [gbCore pushGBButton:PVGBButtonSelect];
	[super pressSelectForPlayer:player];
}

- (void)releaseSelectForPlayer:(NSUInteger)player
{
    PVGBEmulatorCore *gbCore = (PVGBEmulatorCore *)self.emulatorCore;
    [gbCore releaseGBButton:PVGBButtonSelect];
	[super releaseSelectForPlayer:player];
}

@end

File: Provenance/Controller/PVGBControllerViewController.swift
//
//  PVGBControllerViewController.swift
//  Provenance
//
//  Created by Joe Mattiello on 17/03/2018.
//  Copyright (c) 2018 Joe Mattiello. All rights reserved.
//

import PVSupport

fileprivate extension JSButton {
    var buttonTag: PVGBButton {
        get {
            return PVGBButton(rawValue: tag)!
        }
        set {
            tag = newValue.rawValue
        }
    }
}

class PVGBControllerViewController: PVControllerViewController<PVGBSystemResponderClient> {

    override func layoutViews() {
        buttonGroup?.subviews.forEach {
            guard let button = $0 as? JSButton else {
                return
            }
            if (button.titleLabel?.text == "A") {
                button.buttonTag = .a
            } else if (button.titleLabel?.text == "B") {
                button.buttonTag = .b
            }
        }

        startButton?.buttonTag = .start
        selectButton?.buttonTag = .select
    }

    override func dPad(_ dPad: JSDPad, didPress direction: JSDPadDirection) {
        emulatorCore.didRelease(.left, forPlayer: 0)
        emulatorCore.didRelease(.right, forPlayer: 0)
        emulatorCore.didRelease(.up, forPlayer: 0)
        emulatorCore.didRelease(.down, forPlayer: 0)
        switch direction {
            case .upLeft:
                emulatorCore.didPush(.up, forPlayer: 0)
                emulatorCore.didPush(.left, forPlayer: 0)
            case .up:
                emulatorCore.didPush(.up, forPlayer: 0)
            case .upRight:
                emulatorCore.didPush(.up, forPlayer: 0)
                emulatorCore.didPush(.right, forPlayer: 0)
            case .left:
                emulatorCore.didPush(.left, forPlayer: 0)
            case .right:
                emulatorCore.didPush(.right, forPlayer: 0)
            case .downLeft:
                emulatorCore.didPush(.down, forPlayer: 0)
                emulatorCore.didPush(.left, forPlayer: 0)
            case .down:
                emulatorCore.didPush(.down, forPlayer: 0)
            case .downRight:
                emulatorCore.didPush(.down, forPlayer: 0)
                emulatorCore.didPush(.right, forPlayer: 0)
            default:
                break
        }
        super.dPad(dPad, didPress: direction)
    }

    override func dPadDidReleaseDirection(_ dPad: JSDPad) {
        emulatorCore.didRelease(.up, forPlayer: 0)
        emulatorCore.didRelease(.down, forPlayer: 0)
        emulatorCore.didRelease(.left, forPlayer: 0)
        emulatorCore.didRelease(.right, forPlayer: 0)
        super.dPadDidReleaseDirection(dPad)
    }

    override func buttonPressed(_ button: JSButton) {
        emulatorCore.didPush(button.buttonTag, forPlayer: 0)
        super.buttonPressed(button)
    }

    override func buttonReleased(_ button: JSButton) {
        emulatorCore.didRelease(button.buttonTag, forPlayer: 0)
        super.buttonReleased(button)
    }

    override func pressStart(forPlayer player: Int) {
        emulatorCore.didPush(.start, forPlayer: player)
        super.pressStart(forPlayer: player)
    }

    override func releaseStart(forPlayer player: Int) {
        emulatorCore.didRelease(.start, forPlayer: player)
        super.releaseStart(forPlayer: player)
    }

    override func pressSelect(forPlayer player: Int) {
        emulatorCore.didPush(.select, forPlayer: player)
        super.pressSelect(forPlayer: player)
    }

    override func releaseSelect(forPlayer player: Int) {
        emulatorCore.didRelease(.select, forPlayer: player)
        super.releaseSelect(forPlayer: player)
    }
}
